packages/firestore/src/remote/watch_change.ts

```diff
@@ -409,16 +411,81 @@ export class WatchChangeAggregator {
         }
       } else {
         const currentSize = this.getCurrentDocumentCountForTarget(targetId);
+        // Existence filter mismatch. Remove documents to limbo, and raise a
+        // snapshot with `isFromCache:true`.
         if (currentSize !== expectedCount) {
-          // Existence filter mismatch: We reset the mapping and raise a new
-          // snapshot with `isFromCache:true`.
-          this.resetTarget(targetId);
-          this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          // Apply bloom filter to identify and move removed documents to limbo.
+          const bloomFilterApplied = this.applyBloomFilter(
+            watchChange.existenceFilter,
+            targetId,
+            currentSize
+          );
+          if (!bloomFilterApplied) {
+            // If bloom filter application fails, we reset the mapping and trigger
+            // re-run of the query.
+            this.resetTarget(targetId);
+            this.pendingTargetResets = this.pendingTargetResets.add(targetId);
+          }
         }
       }
     }
   }
 
+  /** Returns wheather a bloom filter removed the deleted documents successfully. */
+  private applyBloomFilter(
+    existenceFilter: ExistenceFilter,
+    targetId: number,
+    currentCount: number
+  ): boolean {
+    const unchangedNames = existenceFilter.unchangedNames;
+    const expectedCount = existenceFilter.count;
+
+    if (!unchangedNames || !unchangedNames.bits) {
+      return false;
+    }
+
+    const {
+      bits: { bitmap = '', padding = 0 },
+      hashCount = 0
+    } = unchangedNames;
+    const normalizedBitmap = normalizeByteString(bitmap).toUint8Array();
+    const bloomFilter = new BloomFilter(normalizedBitmap, padding, hashCount);
+
+    if (bloomFilter.size === 0) {
+      return false;
+    }
+
+    const removedDocumentCount = this.filterRemovedDocuments(
+      bloomFilter,
+      targetId
+    );
+
+    if (currentCount - removedDocumentCount === expectedCount) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Filter out removed documents based on bloom filter membership result and return number
+   * of documents removed.
+   */
+  private filterRemovedDocuments(
+    bloomFilter: BloomFilter,
+    targetId: number
+  ): number {
+    const existingKeys = this.metadataProvider.getRemoteKeysForTarget(targetId);
+    let removalCount = 0;
+    existingKeys.forEach(key => {
+      if (!bloomFilter.mightContain(key.path.toFullPath())) {
+        this.removeDocumentFromTarget(targetId, key, /*updatedDocument=*/ null);
```

Comment:
Does this trigger limbo resolution? Obviously, limbo resolution is triggered somehow, but I just don't see where it is happening.

Reply:
Discussed offline.